### ISSUE TITLE
Remove dependency on conda and standardize on poetry

### ISSUE THREAD

**Issue by wherka-ama** [OpenDevin/OpenDevin#1274]
## Describe your question

As you know conda is not quite free anymore. It's been like that for a couple of years now. They changed the licenses. 
The only channel which is still free is the `forge`.  However, I would exercise caution and state it very clearly that it may be problematic for some people who would like to use OpenDevin in their corporate context and they do not wish to pay for the subscriptions as they may be in the breach of the license.

## Additional context

Please read their T&C very carefully and reflect it somewhere in the docs, so it's clear for anyone.
Here is the relevant [section](https://legal.anaconda.com/policies/en/?name=terms-of-service#anaconda-terms-of-service)

BTW: Great project and a very impressive growth of this community! Well done :+1: All the best!

**Comment by Umpire2018** [comment]
Looks like in https://github.com/OpenDevin/OpenDevin/blob/main/evaluation/SWE-bench/environment.yml, opendevin SWE-bench is using `conda-forge`.

**Comment by wherka-ama** [comment]
> Looks like in https://github.com/OpenDevin/OpenDevin/blob/main/evaluation/SWE-bench/environment.yml, opendevin SWE-bench is using `conda-forge`.

It appears that it is indeed setting the channel correctly to `conda-forge`.

However, there is still a bit of a problem here. See [all the references](https://github.com/search?q=repo%3AOpenDevin%2FOpenDevin%20conda&type=code) of conda in the project. They point at the `conda/miniconda` which may be downloaded from `https://repo.anaconda.com/miniconda` which is already creating a problem.

In fact I've found a proof that it is using the official repo and not the forge one:
https://github.com/OpenDevin/OpenDevin/blob/main/containers/evaluation/Dockerfile#L22

The TOS(Terms Of Service) I mentioned earlier state that anything pulled from the official repos(not maintained by the conda-forge community) are automatically in the scope.

Therefore I would still try to clarify it and point explicitly at [miniforge installation](https://github.com/conda-forge/miniforge).

For me it looks like we may still be in a position to offer something that may breach the Anaconda Inc licenses when used in the corporate context.

**Comment by enyst** [comment]
IMO Anaconda licensing is a bit of a minefield.

> The TOS(Terms Of Service) I mentioned earlier state that anything pulled from the official repos(not maintained by the conda-forge community) are automatically in the scope.

I got curious on the details here, so among others, I spot-checked the installation scripts in the repo from the dockerfile that you show above. The scripts include a license for display. One of them, for MacOS, includes the "ToS" license you pointed out, in full. Curiously, one random script for Linux includes a different license, a BSD derivative with advertising clause and some corporate legal CYA paragraph. However, the Linux script for x86_64 has the license you pointed out again, the one with the 200 users restriction.

(It looks to me like the BSD derivative case may be a leftover from the past.)

This included license is shown to the user when they install miniconda manually. When running with -b, it "assumes that the user agreed".

**Comment by enyst** [comment]
This has been fixed by https://github.com/OpenDevin/OpenDevin/pull/1468.

**Comment by marscod** [comment]
> This has been fixed by [#1468](https://github.com/OpenHands/OpenHands/pull/1468).

@enyst and @xingyaoww, I'm wondering how this issue has been addressed. In the latest version, still I'm seeing a reference to `anaconda.org` (despite it is `conda-forge` channel) in following line:
https://github.com/OpenHands/OpenHands/blob/b5e00f577c408867c23a92e61c5b323018828ba0/openhands/runtime/utils/runtime_templates/Dockerfile.j2#L291

```bash
6.919     Download error (35) SSL connect error [https://conda.anaconda.org/conda-forge/noarch/repodata.json.zst]
```
Do I need to set any environment variable or configuration to avoid hitting `anaconda.org` for `conda-forge`, e.g., to use a mirror?

Thank you

**Comment by enyst** [comment]
@openhands Read all this issue comments carefully. You could understand also how it was fixed in PR 1468. (note that it's a large unrelated PR, you don't need most of its diff content)

Then investigate @marscod 's question: why is it hitting anaconda.org?
I see in https://github.com/OpenHands/OpenHands/blob/main/openhands/runtime/utils/runtime_templates/Dockerfile.j2#L290

that we use `micromamba` and in the lines above line 290, a line removes channel defaults.

Intention was always to not install from anaconda.org, simply.

Fetch updated information you need from the web, if you need it. Post your response in this issue, issue 1274, introducing yourself on the first line of the response.

**Comment by enyst** [comment]
Hi, I'm openhands, the OpenHands agent. I dug into why builds still contact anaconda.org.

Short answer: even though we removed the “defaults” channel and explicitly use conda-forge via micromamba, the name conda-forge resolves to a repository hosted on anaconda.org. That’s why you still see requests to anaconda.org, without pulling from the Anaconda “defaults” channel.

Details
- In the Dockerfile template, we:
  - Install micromamba and remove the defaults channel: install.sh with CONDA_FORGE_YES=yes, then `micromamba config remove channels defaults`.
  - Install into the env with `-c conda-forge`:
    /openhands/micromamba/bin/micromamba install -n openhands -c conda-forge poetry python=3.12 -y
- By design, the channel name “conda-forge” is expanded using conda/mamba’s channel_alias, which defaults to the Anaconda domain. So `-c conda-forge` becomes https://conda.anaconda.org/conda-forge and indexes/artifacts are fetched from that host.

References
- Anaconda docs (Channels): the default channel alias is the Anaconda domain, and “the conda-forge channel is hosted at anaconda.org/conda-forge”.
  https://www.anaconda.com/docs/getting-started/working-with-conda/channels
- Upstream discussion confirming conda-forge is served from anaconda.org, and options to avoid it (e.g., using prefix.dev mirror):
  https://github.com/conda-forge/miniforge/issues/784

What PR 1468 accomplished
- It eliminated reliance on Anaconda’s defaults and standardized on using micromamba + poetry. That means we are not pulling packages from the “defaults” channel (repo.anaconda.com/main, etc.).
- However, since conda-forge’s canonical hosting is on the anaconda.org domain, network traffic still goes to anaconda.org when resolving/installing from conda-forge.

Why you saw the SSL error against anaconda.org
- The log line (Download error (35) SSL connect error … anaconda.org/conda-forge/...) is consistent with corporate networks that block the anaconda.org domain. Even when only using conda-forge, you’ll still hit anaconda.org because that’s where conda-forge is hosted by default.

If the intention is “do not hit anaconda.org at all”
- You need to point conda/micromamba at a mirror that is not on anaconda.org. Two common approaches:
  1) Use prefix.dev’s mirror explicitly when installing:
     `/openhands/micromamba/bin/micromamba install -n openhands -c https://repo.prefix.dev/conda-forge poetry python=3.12 -y`
  2) Or set the channel alias before any installs so plain `-c conda-forge` resolves away from anaconda.org:
     `/openhands/micromamba/bin/micromamba config set channel_alias https://repo.prefix.dev`
     `# then keep using: -c conda-forge`

- Enterprises can also front this with an internal proxy (e.g., Artifactory) that mirrors conda-forge and avoids any direct calls to anaconda.org.

Bottom line
- Current code does not use the Anaconda “defaults” channel; it uses conda-forge via micromamba as intended.
- The reason you still see traffic to anaconda.org is that conda-forge is hosted under that domain by default. Avoiding any contact with anaconda.org requires using a different channel URL or an alternate channel_alias (e.g., repo.prefix.dev) or a corporate mirror.

If you’d like, I can open a small follow-up PR to make the channel host configurable in the Dockerfile template (e.g., a Jinja arg like channel_alias with a default), and update docs to mention the anaconda.org hosting and the mirror option.

**Comment by enyst** [comment]
@marscod (myself, not the agent)
If I understand the situation correctly, your blocking any `anaconda`, right? It seems our template, as it currently is, doesn't actually use anaconda default channels. The default hosting for conda-forge is on some anaconda.org domain, though, so if you're blocking everything you're blocking that too.

I found the linked discussion (the miniforge3 link) very interesting.

And this: https://prefix.dev/blog/towards_a_vendor_lock_in_free_conda_experience

**Comment by marscod** [comment]
@enyst, thank you for the quick response.
That’s correct; I’m trying to avoid accessing anaconda.org, but the conda-forge channel still points there (the [same issue](https://github.com/conda-forge/miniforge/issues/784) mentioned by @openhands-agent). I also tried using `https://repo.prefix.dev/conda-forge`, but it seems the OpenHands mirror isn’t available at that location.
```
# Create environment and install packages using the prefix.dev mirror
RUN conda create -n openhands -y -c https://repo.prefix.dev/conda-forge --override-channels && \
    conda install -n openhands -y -c https://repo.prefix.dev/conda-forge --override-channels poetry python=3.12
```

I see that miniforge3 was used previously [here](https://github.com/OpenHands/OpenHands/issues/4153) but there was a blocker. 

> If you’d like, I can open a small follow-up PR to make the channel host configurable in the Dockerfile template (e.g., a Jinja arg like channel_alias with a default), and update docs to mention the anaconda.org hosting and the mirror option.

I like @openhands-agent 's idea.

**Comment by enyst** [comment]
Hi, I'm @openhands, the project's automated assistant.

Why it hits anaconda.org
- We use micromamba and install packages with "-c conda-forge". Even with the defaults channel removed, conda/mamba resolve named channels (like conda-forge) via a configurable base called channel_alias.
- By default, channel_alias points to https://conda.anaconda.org, so "-c conda-forge" expands to https://conda.anaconda.org/conda-forge — which is under anaconda.org. That’s why you see requests there. Removing "defaults" avoids Anaconda’s default repo, but it doesn’t change the host used for conda-forge.

What I changed
- I added a way to override the channel host. If you set an environment variable before building the runtime image, the Dockerfile will run: 
  micromamba config set channel_alias <your-host>
- We still specify "-c conda-forge" for packages, but now it resolves under your chosen host instead of anaconda.org.

How to avoid anaconda.org
Option A: Use prebuilt runtime images
- No conda network traffic occurs on your side during normal use, so no contact with anaconda.org.

Option B: Build the runtime image yourself and set a different host
- Before running your usual build/run that triggers the runtime build, set:

  export OH_CONDA_CHANNEL_ALIAS=https://repo.prefix.dev

- The generated Dockerfile will include: 

  micromamba config set channel_alias https://repo.prefix.dev

- You can also point channel_alias to an internal corporate mirror/proxy (e.g., Artifactory) fronting conda-forge.

PRs
- Implementation (draft): https://github.com/OpenHands/OpenHands/pull/11516
- Docs (draft): https://github.com/OpenHands/docs/pull/57 (new page explains prebuilt vs self-built flows and channel_alias usage)

Notes
- This keeps backward compatibility: if OH_CONDA_CHANNEL_ALIAS is unset, behavior is unchanged.
- A unit test verifies the Dockerfile renders the "micromamba config set channel_alias" line when the env var is set, and that "-c conda-forge" remains.

References
- Conda channels/channel_alias: https://www.anaconda.com/docs/getting-started/working-with-conda/channels
- Background on conda-forge hosting: https://github.com/conda-forge/miniforge/issues/784

If you’d like me to point channel_alias somewhere else by default or cover additional mirrors in docs, I can update the PRs accordingly.

**Comment by enyst** [comment]
Hi, I’m openhands (AI assistant) – following up on #1274 and @marscod’s question about requests to anaconda.org.

Why builds still contact anaconda.org
- Even when using only conda-forge and removing the defaults channel, conda/mamba resolve short channel names via channel_alias. By default, channel_alias is https://conda.anaconda.org, and conda-forge is hosted there. So installs like -c conda-forge expand to https://conda.anaconda.org/conda-forge/…, which is why logs show connections to anaconda.org.
  - Conda docs: The default channel_alias is https://conda.anaconda.org (see “Settings — channel_alias”).

What we changed
- We already removed the defaults channel and use micromamba in the Dockerfile template. To fully avoid hitting anaconda.org, I’ve added a way to override channel_alias at build time.
- New behavior (opt-in): if you set OH_CONDA_CHANNEL_ALIAS, the template will run:
  micromamba config set channel_alias '<your-host>'
  before any installs. We still install packages from -c conda-forge by name; this simply changes the host used to resolve that channel.

How to avoid contacting anaconda.org
- If you build the runtime image locally/CI:
  export OH_CONDA_CHANNEL_ALIAS=https://repo.prefix.dev
  # or point to a corporate mirror/proxy that mirrors conda-forge
  Then run your usual OpenHands build that triggers the runtime image build.
- If you prefer prebuilt images: continue using them. If your environment prohibits any contact with anaconda.org, build your own image with OH_CONDA_CHANNEL_ALIAS set and configure OpenHands to use that custom image.
- Notes:
  - We explicitly remove the defaults channel in the template to prevent fallback to Anaconda’s default repos.
  - channel_alias controls the hostname used for channel resolution – that’s what stops any anaconda.org traffic when set to a different host.

References
- Core PR (draft): https://github.com/OpenHands/OpenHands/pull/11516
  - Fixes: configurable channel_alias in Dockerfile.j2, safe quoting, correct RUN continuation, and unit tests for template rendering.
- Docs PR (draft): https://github.com/OpenHands/docs/pull/57
  - Explains using prebuilt images vs building locally, and shows how setting channel_alias avoids contacting anaconda.org.
- Conda docs on channels/channel_alias: https://docs.conda.io/projects/conda/en/stable/user-guide/configuration/settings.html#channel-alias

Answering @marscod directly
- You’re seeing anaconda.org because -c conda-forge resolves via the default channel_alias (https://conda.anaconda.org). Our change makes that host configurable so you can point it at an alternative (e.g., https://repo.prefix.dev or your internal mirror) and avoid contacting anaconda.org entirely.

Next steps
- I’ve addressed AI review feedback in the template (quoting and RUN continuation) and added tests to validate rendering. Please let me know if you’d prefer a Docker ARG instead of an environment variable; I can update the PR accordingly.
